packages/navigator/src/browser/navigator-widget.tsx

```diff
@@ -88,21 +93,55 @@ export class FileNavigatorWidget extends FileTreeWidget {
                         this.model.expandNode(child);
                     }
                 }
+            }),
+            this.workspaceService.onWorkspaceChanged(async event => {
+                await this.refreshWidget(event.roots, event.workspaceConfigFile);
             })
         ]);
     }
 
-    protected initialize(): void {
-        this.workspaceService.root.then(async resolvedRoot => {
-            if (resolvedRoot) {
-                const uri = new URI(resolvedRoot.uri);
+    protected async initialize(): Promise<void> {
+        const workspaceDescriptionData = await this.workspaceService.workspace;
+        await this.refreshWidget(workspaceDescriptionData.roots, workspaceDescriptionData.workspaceConfigFile);
+    }
+
+    protected async refreshWidget(roots: FileStat[], workspaceConfig: FileStat | undefined): Promise<void> {
+        if (roots.length > 0) {
+            if (roots.length === 1) { // only one folder is opened
+                const root = roots[0];
+                this.model.hasMultipleRoots = false;
+                const uri = new URI(root.uri);
                 const label = this.labelProvider.getName(uri);
-                const icon = await this.labelProvider.getIcon(resolvedRoot);
-                this.model.root = DirNode.createRoot(resolvedRoot, label, icon);
-            } else {
-                this.update();
+                const icon = await this.labelProvider.getIcon(root);
+                this.model.root = DirNode.createRoot(root, label, icon);
+            } else if (roots.length > 1) { // more than one folder
+                if (workspaceConfig) {
+                    this.model.hasMultipleRoots = true;
+                    const workspaceRootNode = {
+                        id: workspaceConfig.uri,
+                        name: WorkspaceRootNode.parentName,
+                        parent: undefined
+                    } as CompositeTreeNode;
+
+                    const children: TreeNode[] = [];
+                    for (const r of roots) {
+                        const icon = await this.labelProvider.getIcon(r);
+                        children.push({
+                            id: r.uri,
+                            name: this.labelProvider.getName(new URI(r.uri)),
+                            icon,
+                            visible: true,
+                            parent: workspaceRootNode
+                        });
+                    }
+                    workspaceConfig.children = roots;
+                    this.model.root = this.createWorkspaceRoot(workspaceConfig, children);
+                }
             }
-        });
+        } else {
+            this.model.root = undefined;
+            this.update();
```

Comment:
calling `update` will cause unnecessray rendering, whenever a tree model is changed a widget is rerendered